### DISCUSSION TITLE
Develop eepp applications with ecode

### DISCUSSION THREAD

**Discussion by ghost** [SpartanJ/eepp#72] [General]
I need code completion by clangd. This requires compile_commands.json. But premake can't generate this file for me. I don't want to use Bear. I'm on Windows 7 (an outdated Windows that support has been dropped by the MSYS2 project). Add more tools to the stack will only causing more problems.

My setup is simple: llvm-mingw (includes clangd) + eepp + ecode. No other external dependencies. It seems you need to add a CMakeLists.txt for this example so I could use it as the foundation for my projects:

https://github.com/SpartanJ/eepp/tree/develop/src/examples/empty_window

CMake is the most portable and the easiest to get on Windows that can generate compile_commands.json.

p/s: I'm stuck with cmake 3.28.6 as it is the last version supports Windows 7.

**Comment by SpartanJ** [comment]
Casually last night I was exposing the export compile commands functionality in eepp repository (commited [here](https://github.com/SpartanJ/eepp/commit/46b2a1b7692c898d05eeaaf4cb11c3ff152bb0e5)). Premake5 does support this but you need to install a module, now the module will be integrated in eepp, so you only need to run:
`premake5 export-compile-commands ` and it will generate the `compile_commands.json` file in the root directory of the project.
Premake5 is quite flexible, there are modules for generating [cmake](https://github.com/Jarod42/premake-cmake) and [ninja](https://github.com/jimon/premake-ninja) projects. [Here](https://premake.github.io/community/modules/) there is a list of some of the modules available.

If you need to add your own projects by extending the premake5 configuration of eepp it's really easy since i did that to easily compile [eeiv](https://github.com/SpartanJ/eeiv). `premake5.lua` file will also include (if exists) an `external_projects.lua` file to generate external project applications (like what you're requiring).

For example my local `external_projects.lua` looks like:
```lua
	project "eeiv"
		kind "WindowedApp"
		language "C++"
		files { "../eeiv/src/*.cpp" }
		build_link_configuration( "eeiv", true )
```

This will link the project against eepp, build the project in the and `bin` folder. To build that particular project you just need to make the name of the project, for example:
`mingw32-make -C make\\windows eeiv`

And that's all you need.

**Comment by ghost** [comment]
https://github.com/tarruda/premake-export-compile-commands

This project has not been updated for 9 years. And `clangd` looks for a file named `compile_commands.json`. This file is expected to be in the root source tree or in the `build` directory. It will not see and load the json files in `compile_commands` directory.

I found this project on the Premake modules page:

https://github.com/MattBystrin/premake-ecc

Maybe it's a better candidate.

**Reply by ghost** [reply]
@Jarod42 has a fork with some improvements regarding this.

**Reply by ghost** [reply]
In his fork, he renamed the json files to the correct name, `compile_commands.json`, and also changed the locations of them. But `clangd` still can't find the `compile_commands.json` file. As I said, `clangd` expects this file to be in the root source tree or in the `build` directory.

**Reply by Jarod42** [reply]
From https://clangd.llvm.org/installation.html
>  If your build directory is $SRC or $SRC/build, clangd will find it. Otherwise, symlink or copy it to $SRC, the root of your source tree.
> 
> `ln -s ~/myproject-build/compile_commands.json ~/myproject/`

**Reply by SpartanJ** [reply]
@bxijt4 I don't think you tried to use what comes with eepp. I forked @Jarod42 version and made a small modification: my version will just check your arch an copy the file to the root of the repository, check it [here](https://github.com/SpartanJ/eepp/blob/develop/premake/export-compile-commands/export-compile-commands.lua#L190), it's only useful for eepp since it's getting the release version following the project convention.

**Comment by ghost** [comment]
@SpartanJ Does `external_projects.lua` work with `premake4`? How a `external_projects.lua` for `premake4` look like? I don't think I can reuse the `external_projects.lua` for `premake5`. Debian 12 only has `premake4` (#53).

**Reply by SpartanJ** [reply]
> Does external_projects.lua work with premake4?

The eeiv example works in both premake(4|5) versions.

> How a external_projects.lua for premake4 look like?

It depends on your needs, for a trivial case you can share the project in both premakes, given that `build_link_configuration` function is available in both premake projects.

**Comment by ghost** [comment]
@Jarod42 @SpartanJ Can you tell me what's going on with `premake5 export-compile-commands`? I'm using GCC and don't have Clang. I use the standalone Clangd. But why the compile commands are like this? Why does it think I'm using Clang as the compiler?

```
	{
		"directory": "C:/Work/eepp/make/windows",
		"file": "C:/Work/eepp/src/thirdparty/SOIL2/src/SOIL2/SOIL2.c",
		"command": "clang -x c -MD -MP -IC:/Work/eepp/src/thirdparty/SOIL2 -IC:/Work/eepp/src/thirdparty/zlib -m64 -O3 -std=gnu99 -Wall -o \"C:/Work/eepp/obj/windows/x86_64/release/SOIL2-static/SOIL2.o\" -MF \"C:/Work/eepp/obj/windows/x86_64/release/SOIL2-static/SOIL2.d\" -c \"C:/Work/eepp/src/thirdparty/SOIL2/src/SOIL2/SOIL2.c\""
	},
```

**Reply by Jarod42** [reply]
that Action defaults `toolset` to `clang`
```
newaction {
  trigger = 'export-compile-commands',
  description = 'Export compiler commands in JSON Compilation Database Format',
  onWorkspace = m.onWorkspace,
  toolset = "clang",
 ```

Not sure if it has an effect to have `clang` or `gcc` or `cc` at that place.
Clang tooling parses the **options**, mainly include path and defines.

`premake5 export-compile-commands --cc=gcc` should display `gcc` instead.
